cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/services/FlowControlMonitorService.java

```diff
@@ -0,0 +1,189 @@
+import io.cdap.cdap.proto.ProgramRunStatus;
+import io.cdap.cdap.proto.ProgramType;
+import io.cdap.cdap.proto.id.ProgramRunId;
+import io.cdap.cdap.spi.data.transaction.TransactionRunner;
+import io.cdap.cdap.spi.data.transaction.TransactionRunners;
+import java.util.Collections;
+import java.util.EnumSet;
+import java.util.Set;
+import java.util.concurrent.TimeUnit;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Maintain and provides total number of launching and running run-records. This class is used by
+ * flow-control mechanism for launch requests.
+ */
+public class FlowControlMonitorService extends AbstractIdleService {
+
+  private static final Logger LOG = LoggerFactory.getLogger(FlowControlMonitorService.class);
+
+  // Program types controlled by flow-control mechanism.
+  private static final Set<ProgramType> CONTROL_FLOW_PROGRAM_TYPES = EnumSet.of(ProgramType.MAPREDUCE,
+      ProgramType.WORKFLOW,
+      ProgramType.SPARK,
+      ProgramType.WORKER);
+
+  private final MetricsCollectionService metricsCollectionService;
+  private final TransactionRunner transactionRunner;
+
+  /**
+   * Monitors the program flow control.
+   *
+   * @param metricsCollectionService collect metrics
+   */
+  @Inject
+  public FlowControlMonitorService(
+      MetricsCollectionService metricsCollectionService,
+      TransactionRunner transactionRunner) {
+    this.metricsCollectionService = metricsCollectionService;
+    this.transactionRunner = transactionRunner;
+  }
+
+  @Override
+  protected void startUp() throws Exception {
+    LOG.info("FlowControlMonitorService started.");
+  }
+
+  @Override
+  protected void shutDown() throws Exception {
+    LOG.info("FlowControlMonitorService successfully shut down.");
+  }
+
+  /**
+   * Add a new in-flight launch request and return total number of launching and running programs.
+   *
+   * @param programRunId run id associated with the launch request
+   * @return total number of launching and running program runs.
+   */
+  public Counter addRequestAndGetCount(ProgramRunId programRunId, ProgramOptions programOptions,
+      ProgramDescriptor programDescriptor) throws Exception {
+    if (RunIds.getTime(programRunId.getRun(), TimeUnit.MILLISECONDS) == -1) {
+      throw new Exception("None time-based UUIDs are not supported");
+    }
+
+    int launchingCount = addRequest(programRunId, programOptions, programDescriptor);
+    int runningCount = getFlowControlMetrics(false, true).getRunningCount();
+
+    LOG.info(
+        "Counter has {} concurrent launching and {} running programs.",
+        launchingCount,
+        runningCount);
+    return new Counter(launchingCount, runningCount);
+  }
+
+  /**
+   * Get total number of launching and running programs.
+   *
+   * @return Counter with total number of launching and running program runs.
+   */
+  public Counter getCount() {
+    return getFlowControlMetrics(true, true);
```

Comment:
Let's not use term metrics as it's pretty strongly associated with metrics service. Use counts.

Reply:
Renamed with `Counter`.